### ISSUE TITLE
Error: "Couldn't find the word ... in the text"

### ISSUE THREAD

**Issue by smoores-dev** [echogarden-project/echogarden#67]
This just popped up for me for the first time. Running the `recognize` function (with whisper.cpp, built with OpenBLAS, on CPU) on what is, as far as I know, not a pathological audio sample (it's an audiobook narration) resulted in this error:

```
Error: Couldn't find the word 'I...I' in the text at start position 8544
    at wordTimelineToSegmentSentenceTimeline (file:///app/node_modules/echogarden/dist/utilities/Timeline.js:88:19)
    at recognize (file:///app/node_modules/echogarden/dist/api/Recognition.js:184:39)
    at async transcribeTrack (file:///app/web/work-dist/worker.js:2190:70)
    at async transcribeBook (file:///app/web/work-dist/worker.js:2310:29)
    at async processBook (file:///app/web/work-dist/worker.js:2404:9)
    at async /app/node_modules/piscina/dist/src/worker.js:145:26
```

If there's a way for me to poke around further and provide more information, please let me know!

**Comment by smoores-dev** [comment]
Huh, oddly, it worked fine on the same file with whisper.cpp not using OpenBLAS. It's possible that the OpenBLAS build is somehow providing buggy output.

**Comment by hhu94** [comment]
I got the exact same error (running Storyteller 😉), but built with cuBLAS 12.4.

> 2024-09-03 00:21:27 web-1  | Error: Couldn't find the word 'I...I' in the text at start position 22613
2024-09-03 00:21:27 web-1  |     at wordTimelineToSegmentSentenceTimeline (file:///app/node_modules/echogarden/dist/utilities/Timeline.js:88:19)
2024-09-03 00:21:27 web-1  |     at recognize (file:///app/node_modules/echogarden/dist/api/Recognition.js:184:39)
2024-09-03 00:21:27 web-1  |     at async transcribeTrack (file:///app/web/work-dist/worker.js:2204:70)
2024-09-03 00:21:27 web-1  |     at async transcribeBook (file:///app/web/work-dist/worker.js:2324:29)
2024-09-03 00:21:27 web-1  |     at async processBook (file:///app/web/work-dist/worker.js:2422:9)
2024-09-03 00:21:27 web-1  |     at async /app/node_modules/piscina/dist/src/worker.js:145:26

**Comment by smoores-dev** [comment]
It's wild that this seems specific to "I...I". Someone else just reported this on Reddit, and they were seeing it with "I...I...I". Feels like a whisper bug, but I'm going to try to dig into it tomorrow and see if I can pinpoint the issue.

**Comment by smoores-dev** [comment]
... dang. I don't seem to be able to reproduce this on my development machine, even with the same configuration (using the same whisper.cpp build and model)

**Comment by rotemdan** [comment]
The error is similar to #58 (you can see the explanation there). 

It occurs whenever a recognizer parses the words of the text differently than the word and sentence segmentation produced by the `cldr-segmentation` library. Usually the problem is when there are dots (`.`) in the middle of the word, which `cldr-segmentation` usually interprets as the end of a sentence, so it splits it in the middle. 

Because the recognizer's segmentation is different than the one used to identify sentence boundaries, it can't match the word the recognizer returns in the other segmentation.

In this case the recognizer produces some sort of a pattern of letters and dots, like `I...I` and treats it as a single word because of the whitespace around it. 

I need to find a way to work around these kinds of problems and not just produce an error. I had this on my list for at least 6 months now.

**Comment by smoores-dev** [comment]
Ah, ok, I was just poking around the cldr-segmentation/sentence splitting code, so I guess good to know that I was on the right track! Thanks for the update @rotemdan

**Comment by rotemdan** [comment]
Usually the order of operations is that I first segment to sentences, and then segment those sentences, individually to words. That wouldn't work in these edge cases. 

I could split recognized words that contain dots, but then it also means that valid ones like `U.S.A` would also be split, and that would force the sentence segmentation to split it to 3 sentences `U.`, `S.`, `A.`. That a part of makes this problem difficult to fix.

I could also maybe first split to words, then temporarily hide all dots in the body of those words (dots that are outside will remain as individual words), and then merge those words to text again, and then split to sentences, and use that sentence segmentation to categorize the already segmented words to sentences. Something like that maybe. I'll need to see.

**Comment by smoores-dev** [comment]
Can you segment to words, and then use the resulting words as suppressions for cldr (maybe ignoring boundary punctuation)? That is, something like:

```
"I...I don't know"
# to words:
["I...I", "don't", "know."]
# trim boundary punctuation
["I...I", "don't", "know"]
# use as custom suppressions
sentenceSplit(
  text,
  suppressions.en.merge(
    Suppressions.create(
      ["I...I", "don't", "know"]
    )
  )
)
```

**Comment by smoores-dev** [comment]
Oh... wait. I think I was understanding the problem backwards. Here's what I'm seeing happen for whisper.cpp, at least:

1. Whisper runs and outputs a token timeline
2. We convert that token timeline to a word timeline, basically by concatenating tokens
3. We also convert the token timeline to a full transcript (a single string)
4. We pass the word timeline to `wordTimelineToSegmentSentenceTimeline`
5. We take the full transcript and attempt to split it into first paragraphs/segments, and then sentences
6. We produce a _new_ transcript (`text`), which is the result of splitting the full transcript into sentences, _trimming them_, and then joining them with a single space character between each
7. We iterate through the word timeline and attempt to find each word in the _new_ transcript

Step 7 is where we run into this error case, but I think step 6 is actually where the problem lies. We assume that each sentence that cldr finds is separated by a space in the original transcript produced by the token timeline, but this isn't necessarily the case. In this specific case, the token timeline produces the transcript `"I...I don't know"`, and cldr produces `["I...", "I don't know"]`. This means that our _new_ transcript, produced from the sentences, reads `"I... I don't know"` (note the space between the ellipsis and the second "I").

I think that one possible solution to this could be to stop trimming the sentences from cldr, and drop the appended `' '` when producing `text`. This doesn't require managing any discrepancies between the cldr segmentation and the word timeline; now `text` will be precisely the same as `transcript` (at least, within a given paragraph), and so searching for the word `"I...I"` within `text` will still work, it's just that it will overlap two sentences. 

The sentence overlap thing may still be an issue? I'm unsure whether it would be a problem to have something like:

```json
[{
  "type": "sentence",
  "text": "I...",
  "words": [{
    "type": "word",
    "text": "I...I",
  }],
},
{
  "type": "sentence",
  "text": "I don't know.",
  "words": [{
    "type": "word",
    "text": "don't",
  },
  {
    "type": "word",
    "text": "know.",
  }]
}]
```

If that would be an issue, then we could attempt to go back and split the word up at the sentence boundary in those scenarios.

**Comment by smoores-dev** [comment]
I guess my other thought here is that Storyteller actually only cares about the word timeline... so maybe it would be cool if there was a way to configure `recognize` to skip the segment timeline calculation step?

**Comment by rotemdan** [comment]
I think in general it could be actually better to first split to words and only then to sentences, since the word segmentation has better handling of abbreviations, if I remember correctly. 

I also noticed that with speech synthesis - sometimes it cuts sentences due to dots in places where the word segmenter would have found a word. It causes a break in the speech.

I think the correct way would be to change how sentence segmentation is done in general, and use a more sophisticated approach that combines it with the language aware word segmenter, to first pick up the language-specific abbreviations, hide them, and only then let the sentence segmenter work on the filtered words.

I prefer to have sentence boundaries on all outputs (even if inaccurate) since it makes the timeline structure more uniform. Otherwise it becomes more complex to parse it since you have to check for each segment if it contains sentences or words.

Anyway, I think most operations already output the word timeline separately in the API. The error is produced when it produces the segment timeline.

**Comment by smoores-dev** [comment]
> Anyway, I think most operations already output the word timeline separately in the API. The error is produced when it produces the segment timeline.

Right, I guess that's what I was asking for; a way to return _just_ the word timeline (which is already returned separately), and skip creating the segment timeline entirely (since it has this issue, and not all consumers will need it). But I understand that that makes the API more dynamic and possibly more confusing, no worries if you're not interested in that approach.

**Comment by smoores-dev** [comment]
@hhu94 I managed to work around this in Storyteller for now! https://gitlab.com/smoores/storyteller/-/releases/web-v0.4.5

**Comment by rotemdan** [comment]
I added a potential workaround for this issue on the new `1.6.0` release. 

The workaround first masks any potential sentence terminator in the body of words before splitting to sentences (I currently use the `_` as the temporary replacement one). Then, it restores the original text.

The source diff is [here](https://github.com/echogarden-project/echogarden/commit/662c7f55001346524ff692c4b607d1a50f3651a6).

Let me know if the issue still persists (likely would be easy to fix at this point) or you encounter any other issues.

**Comment by smoores-dev** [comment]
It's non-trivial to test this (it wasn't entirely deterministic), but I've updated to this version in Storyteller! I'll close this for now, and re-open if it pops up again. Thanks so much for all of your work on these issues @rotemdan!

**Comment by rotemdan** [comment]
Turns out the fix, added in `v1.6.0` had an [incorrect string concatenation](https://github.com/echogarden-project/echogarden/issues/74#issuecomment-2492828218) that prevented it from working correctly. It is fixed in `v2.0.5`.

Anyway, I've developed a new text segmentation library that combines word, phrase and sentence segmentation in a more integrated way. The new library improves over the current one, especially since it directly handles of East-Asian languages like Chinese, Japanese, Thai and Khmer (via second-stage processing with a new WebAssembly port of the [ICU C++ library](https://icu.unicode.org/)), without need for special options or language detection. It also support more special word patterns, and less-common number formats (e.g. spaced separators `2 000 000`).

It will be likely published at `@echogarden/text-segmentation`. (and the ICU segmentation WASM port at `@echogarden/icu-segmentation-wasm`). 

In general, having "in-house" text segmentation library would make it significantly easier for me to work with these kinds of issues. For example, I would remove the entire workaround I did here and simply pass the segmenter a list of predefined word ranges to work with, for phrase and sentence segmentation. That would be so much easier, and probably the "right" way to do things in this case.
